go/internal/services/usagelimiter/redis.go

```
@@ -0,0 +1,380 @@
+		return s.dbFallback.Limit(ctx, req)
+	}
+
+	// If we successfully initialized the key, return the appropriate response
+	if wasSet {
+		if hasSufficientCredits {
+			// Successful decrement - credit change already applied in initValue
+			return s.handleDecrementResult(req, redisKey, initValue)
+		} else {
+			// Insufficient credits - return denial without decrementing
+			metrics.UsagelimiterDecisions.WithLabelValues("redis", "denied").Inc()
+			return UsageResponse{Valid: false, Remaining: int32(currentCredits)}, nil
```

golangci-lint (2.2.2)

[high] G115: integer overflow conversion int64 -> int32

(gosec)